### PR TITLE
Feature: Code fixes

### LINKED ISSUE #615: Implementation: Suggested Code fixes
From design #497

## Change

- Diagnostic takes a new property `fixes` which is a list of code fixes that can be run.

```ts
reportDiagnostic(program, {
   code: "library-not-found", 
   message: "Cannot find library @cadl-lang/rest",
   target: node,
   codefixes: [
     createInstallLibCodeFix("@cadl-lang/rest"),
   ]
})

export interface CodeFix {
  id: string;
  label: string;
  fix: (fixContext: FixContext) => CodeFixAction | CodeFixAction[] | Promise<void> | void
}

export interface FixContext {
  replaceText(node: Node, newText: string): CodeAction;
}

export function createInstallLibCodeFix(library: string) {
  return {
    id: "install-lib",
    label: "Install library",

  }
}
```

Rules that could have a simple fixer:
- `no-enum` https://github.com/Azure/typespec-azure/pull/3915

### PR DESCRIPTION
**Pull request by timotheeguerin**
closes #615

### Suppress
Kapture 2024-02-05 at 15 16 22

### `number` -> `float64` typo fix

Kapture 2024-02-06 at 09 50 28

### Enum to extensible enum in typespec-azure

https://github.com/Azure/typespec-azure/pull/258

### CODE REVIEW

File: packages/compiler/src/core/program.ts [hunk 1]
```diff
@@ -1047,6 +1048,11 @@ export async function compile(
       return;
     }
 
+    if (diagnostic.severity === "warning" && diagnostic.target !== NoTarget) {
```
Comment (markcowl) [inline packages/compiler/src/core/program.ts:1051]: Not sure we want to make it quite this easy to suppress

File: .vscode/launch.json
Comment (timotheeguerin) [inline .vscode/launch.json:1]: revert?

File: packages/compiler/src/testing/rule-tester.ts [hunk 2]
```diff
@@ -42,10 +53,33 @@ export function createLinterRuleTester(
       const diagnostics = await diagnose(code);
       expectDiagnostics(diagnostics, match);
     }
+
+    function applyCodeFix(fixId: string) {
+      return { toEqual };
+
+      async function toEqual(expectedCode: string) {
+        const diagnostics = await diagnose(code);
+        const codefix = diagnostics[0].codefixes?.find((x) => x.id === fixId);
```
Comment (markcowl) [inline packages/compiler/src/testing/rule-tester.ts:62]: This is expected to be used only when there is a single diagnostic?

Comment (timotheeguerin) [reply packages/compiler/src/testing/rule-tester.ts:62]: Yeah the expectation is you emit a single linter warning and apply the code fix on it

Comment (timotheeguerin) [reply packages/compiler/src/testing/rule-tester.ts:62]: If we endup having cases where there is multiple diagnostic always emitted, I think then we can add a extra option to choose which one

**Review by markcowl — APPROVED** [review]

**Pull request merged**